Add migration files to save model usage

Rindrics opened the pull request:
## Summary
<!-- Briefly describe the changes and the purpose of the PR. -->

## Related Issue
<!-- Mention the related issue number if applicable. -->

## Changes
<!-- List the main changes made in this PR. -->

## Testing
<!-- Briefly describe the testing steps or results. -->

## Other Information
<!-- Add any other relevant information for the reviewer. -->

Rindrics commented on apps/studio.giselles.ai/drizzle/schema.ts:
```diff
@@ -333,3 +333,32 @@ export const githubRepositoryEmbeddings = pgTable(
+	createdAt: timestamp("created_at").defaultNow().notNull(),
+});
+
+export type UsageMetric = "input_token" | "output_token";
+
+export type UsageUnit = "tokens";
+
+export const modelUsageItems = pgTable("model_usage_items", {
+	dbId: serial("db_id").primaryKey(),
+	modelUsageDbId: integer("model_usage_db_id")
+		.notNull()
+		.references(() => modelUsage.dbId, { onDelete: "cascade" }),
+	usageMetric: text("usage_metric").$type<UsageMetric>().notNull(),
+	amount: integer("amount").notNull(),
+	unit: text("unit").$type<UsageUnit>().notNull(),
```
I'll leave the code as it is since it can benefit from TypeScript type checking

satococoa reviewed (commented):
Thank you!
I've added a comment. Could you take a look?

Rindrics commented on apps/studio.giselles.ai/drizzle/schema.ts:
```diff
@@ -333,3 +333,46 @@ export const githubRepositoryEmbeddings = pgTable(
+		dbId: serial("db_id").primaryKey(),
+		teamDbId: integer("team_db_id")
+			.notNull()
+			.references(() => teams.dbId, { onDelete: "cascade" }),
```
fixed: 6331e59b

shige commented:
That's a good idea! 
I think it will resolve my concerns because it will be unique. 👍

Rindrics commented:
> > 1\. The combination of `model` and `provider` columns in `model_usage` table directly maps to the keys in our `ModelPriceTable`, allowing us to reference the pricing information.
> 
> Do these keys necessarily match those of `ModelPriceTable`?
> If so, there's no problem, but they could not be read from the existing Table definition.

Thanks, the consistency between these values are not ensured  at that point.
I fixed this problem by referring model names using constant in https://github.com/giselles-ai/giselle/pull/960/commits/d030a257b592f19e723fdc4b07644608a00642bc 👍 

> With the current implementation, the data stored in the database does not meet the minimum requirements, and it is easy to imagine that supporting customers would become very time-consuming.

That's right. It was an important suggestion...
Also applying advice from @satococoa, I added a column `priceId` to the table  `model_usages` （https://github.com/giselles-ai/giselle/pull/960/commits/105e1bbec3743e3f0a710eac7e52aaf21cd3fff8 ）

@shige

shige commented:
Since the scope of the modifications extends to `packages/language-model/src/models.ts`, I would like to ask @toyamarinyon san to review it as well.

toyamarinyon commented on packages/language-model/src/models.ts:
```diff
@@ -0,0 +1,35 @@
+} as const;
+
+export const ANTHROPIC_MODELS = {
+	CLAUDE4_OPUS: "claude-4-opus-20250514",
```
Before discussing the merits of this expression/approach, I want to know why this change is necessary for achieving this Pull request's objective of 'Add DB schema to save model usage'.

Rindrics replied:
After https://github.com/giselles-ai/giselle/pull/960#issuecomment-2915282825, I believe now we can discuss this @toyamarinyon @shige @satococoa 

If we can discuss about the expression of model keys in this PR, I think the key format of `claude-4-opus-20250514-1` is better than the other formats.

Rindrics replied:
I introduced model names as constant to ensure consistency between table column `model` and pricing object stored in `language-models/cost/model-pricing.ts`. @toyamarinyon 

ref: https://github.com/giselles-ai/giselle/pull/960#issuecomment-2911840020
> > > 1\. The combination of `model` and `provider` columns in `model_usage` table directly maps to the keys in our `ModelPriceTable`, allowing us to reference the pricing information.
> > 
> > Do these keys necessarily match those of `ModelPriceTable`?
> > If so, there's no problem, but they could not be read from the existing Table definition.
> 
> Thanks, the consistency between these values are not ensured  at that point.
> I fixed this problem by referring model names using constant in https://github.com/giselles-ai/giselle/pull/960/commits/d030a257b592f19e723fdc4b07644608a00642bc 👍

toyamarinyon reviewed (commented):
At what granularity do you want to record and measure usage? I read the schema, code, and related issues, but I couldn't understand it well.

Rindrics commented:
> At what granularity do you want to record and measure usage?

We have to record tokens (not costs) consumed by each generation, such as the values included in the [response of AI SDK `streamText()`](https://ai-sdk.dev/docs/reference/ai-sdk-core/stream-text#on-step-finish.on-step-finish-result.usage). @toyamarinyon

toyamarinyon commented:
> We have to record tokens (not costs) consumed by each generation, such as the values included in the [response of AI SDK `streamText()`](https://ai-sdk.dev/docs/reference/ai-sdk-core/stream-text#on-step-finish.on-step-finish-result.usage).

@Rindrics  For this use case, I think it would be natural to add a usage object to Generation, but why don't you do that?

Rindrics commented:
Records of model usage should be stored as a separate table because we're going to use them in later billing process @toyamarinyon

Rindrics commented:
Thank you for your feedback. I apologize if [my response to the question about "what granularity we should record"](https://github.com/giselles-ai/giselle/pull/960#issuecomment-2914743247) was misleading. After careful consideration, I believe our current implementation remains the most appropriate for our needs. Let me explain why.

Looking at your proposed changes, I now understand the intent behind the granularity question.
What we ultimately need is "team-based model usage costs (including API calls and image generations in the future) for Stripe billing periods."
When I mentioned that "we need token counts per generation," I didn't mean per generation ID, but rather "per timestamp."
We need timestamps because model pricing might change over time.
While having a reference to generation IDs could be convenient, we don't currently need to investigate usage per specific generation, so our focus remains on "usage" itself.
Therefore, I believe `model_usages` is still a more appropriate table name than generation_usage.

Regarding the "unnecessarily complex" comment, we need to support usage metrics beyond token types (such as API call counts and image metrics) in the future.
This is why we've separated `modelUsageItems` into the separated table and included the `usageMetric` column.

@toyamarinyon

Rindrics commented on apps/studio.giselles.ai/migrations/0034_glorious_owl.sql:
```diff
@@ -0,0 +1,36 @@
+	"provider" text NOT NULL,
+	"agent_db_id" integer NOT NULL,
+	"created_at" timestamp DEFAULT now() NOT NULL
+);
```
Index added by https://github.com/giselles-ai/giselle/pull/960/commits/f76dc6d6c78432eda7728b5576ce62e2d7b4af2b

Rindrics commented on apps/studio.giselles.ai/migrations/0034_glorious_owl.sql:
```diff
@@ -0,0 +1,36 @@
+	"created_at" timestamp DEFAULT now() NOT NULL
+);
+--> statement-breakpoint
+CREATE TABLE IF NOT EXISTS "model_usage_items" (
+	"db_id" serial PRIMARY KEY NOT NULL,
+	"model_usage_db_id" integer NOT NULL,
+	"usage_metric" text NOT NULL,
+	"amount" integer NOT NULL,
+	"unit" text NOT NULL,
+	"ended_at" timestamp NOT NULL,
+	"created_at" timestamp DEFAULT now() NOT NULL
+);
```
I'll ignore this because index added by https://github.com/giselles-ai/giselle/pull/960/commits/f76dc6d6c78432eda7728b5576ce62e2d7b4af2b

Rindrics commented on packages/language-model/src/costs/model-prices.ts:
```diff
@@ -1,322 +1,284 @@
+			price: {
+				input: {
+					costPerMegaToken: 1.1,
+				},
+				output: {
+					costPerMegaToken: 4.4,
 				},
 			},
-		],
-	},
-	"o4-mini": {
-		prices: [
-			{
-				validFrom: "2025-05-12T00:00:00Z",
-				price: {
-					input: {
-						costPerMegaToken: 1.1,
-					},
-					output: {
-						costPerMegaToken: 4.4,
-					},
+		},
+	],
+	[OPENAI_MODELS.O4_MINI]: [
+		{
+			validFrom: "2025-05-12T00:00:00Z",
+			price: {
+				input: {
+					costPerMegaToken: 1.1,
+				},
+				output: {
+					costPerMegaToken: 4.4,
 				},
 			},
-		],
-	},
-};
+		},
+	],
+});
```
The proposed changes below contain hallucination-induced errors and should be disregarded:
```diff
 [OPENAI_MODELS.GPT4O]: [
  {
    validFrom: "2025-05-12T00:00:00Z",
    price: {
-     input: { costPerMegaToken: 2.5 },
-     output: { costPerMegaToken: 10.0 },
+     input: { costPerMegaToken: 3.75 },
+     output: { costPerMegaToken: 15.0 },
    },
  },
 ],
 
 [OPENAI_MODELS.GPT4O_MINI]: [
  {
    validFrom: "2025-05-12T00:00:00Z",
    price: {
-     input: { costPerMegaToken: 0.15 },
-     output: { costPerMegaToken: 0.6 },
+     // TODO: confirm current GPT-4o-mini input/output rates
    },
  },
 ],
```

OpenAI pricing table: https://platform.openai.com/docs/pricing#latest-models

Rindrics replied:
We’re still figuring out the direction for this file [in another conversation](https://github.com/giselles-ai/giselle/pull/960#discussion_r2110614806), so I’ll mark this resolved for now 🙏

Rindrics commented:
Thank you @toyamarinyon 

> **Current Issues:**
- > ❌ Can't calculate costs using SQL queries alone - requires application logic
  - true. I understand this
- > ❌ Price changes require code deployment - business configuration shouldn't need developer intervention  
  - true. This is an important suggestion
- > ❌ Historical pricing is hard to manage - difficult to recalculate costs for past periods
  - false. `priceId` introduced by https://github.com/giselles-ai/giselle/pull/960/commits/105e1bbec3743e3f0a710eac7e52aaf21cd3fff8 enables us to query usage during given period
- > ❌ Non-technical team members can't update pricing - requires code changes
  - true. This is an important suggestion

Although we discussed current design before, would you give me your opinion at this point? 🙏  @satococoa

satococoa commented:
@toyamarinyon cc/ @Rindrics @shige 

I believe storing model pricing data in our codebase is a reasonable approach.

- Pros:
  - Eliminates the need for database pricing management interfaces or scripts
    - I want to avoid manual database operations
  - Price changes are tracked in git history
  - Enables atomic deployment of new models with pricing
  - Pricing changes require deployment (which I consider an advantage)
- Cons:
  - SQL-based calculations become impossible

satococoa commented:
Another idea 💡 

- Having pricing table as a seed data for DB.
- automatically apply the data with deployment.
